presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -506,6 +507,121 @@ private static Domain extractEquatableDomain(ComparisonExpression.Type compariso
             }
         }
 
+        private Optional<Expression> coerceComparisonWithRounding(
+                Type fieldType,
+                QualifiedNameReference fieldReference,
+                Type valueType,
+                Object value,
+                ComparisonExpression.Type comparisonType)
+        {
+            requireNonNull(value, "value is null");
+            Optional<Object> coercedValue = floorValue(valueType, fieldType, value);
+            if (!coercedValue.isPresent()) {
+                return Optional.empty();
+            }
+            Expression result = rewriteComparisonExpression(
+                    fieldType,
+                    fieldReference,
+                    valueType,
+                    value,
+                    coercedValue.get(),
+                    comparisonType
+            );
+            return Optional.of(result);
+        }
+
+        private Expression rewriteComparisonExpression(
+                Type fieldType,
+                QualifiedNameReference fieldReference,
+                Type valueType,
+                Object originalValue,
+                Object coercedValue,
+                ComparisonExpression.Type comparisonType)
+        {
+            int originalComparedToCoerced = compareOriginalValueToCoerced(valueType, originalValue, fieldType, coercedValue);
```

Comment:
given that you are using a saturated cast to do the coercion, i don't think you can do the above comparison if saturation has happened?